File: 2020/day_03/day_03.py
with open('in') as f:
    data = list(map(lambda x: x.strip(), f.readlines()))
h, w = len(data), len(data[0])


def part_one():
    nr_trees = 0
    x, y = 0, 0
    for i in range(h - 1):
        x = (x + 3) % w
        y += 1
        if data[y][x] == '#':
            nr_trees += 1

    return nr_trees


def part_two():
    slopes = ((1, 1), (3, 1), (5, 1), (7, 1), (1, 2))
    res = 1
    for dx, dy in slopes:
        nr_trees = 0
        x, y = 0, 0
        while y + dy < h:
            x = (x + dx) % w
            y += dy
            if data[y][x] == '#':
                nr_trees += 1

        res *= nr_trees

    return res


if __name__ == '__main__':
    print(f'Part One: {part_one()}')
    print(f'Part Two: {part_two()}')

File: 2020/day_03/day_03.rs
use std::fs;

fn main() {
    let f = fs::read_to_string("in").unwrap();
    let data: Vec<Vec<char>> = f.lines().map(|x| x.trim().chars().collect()).collect();

    println!("Part One: {}", part_one(&data));
    println!("Part Two: {}", part_two(&data));
}

fn ride(data: &Vec<Vec<char>>, dx: usize, dy: usize) -> u64 {
    let width = data[0].len();

    let mut count = 0;

    let mut x = 0;
    let mut y = 0;

    while y < data.len() {
        if data[y][x] == '#' {
            count += 1;
        }

        x = (x + dx) % width;
        y += dy
    }

    count
}

fn part_one(data: &Vec<Vec<char>>) -> u64 {
    ride(data, 3, 1)
}

fn part_two(data: &Vec<Vec<char>>) -> u64 {
    let moves = [(1, 1), (3, 1), (5, 1), (7, 1), (1, 2)];
    moves.iter().map(|(dx, dy)| ride(data, *dx, *dy)).product()
}
